### DISCUSSION TITLE
Should the error include the name of the missing variable?

### DISCUSSION THREAD

**Discussion by allan2** [allan2/dotenvy#25] [Ideas]
One of the most common errors is when an env var could not be found.

Currently, the name of the missing env var is not included in the error because `std::env::VarError` is like this:
```rs
enum VarError {
    NotPresent,
    NotUnicode(OsString)
}
```

This means users who want context on the name need to handle the name themselves.
```rs
dotenvy::var("FOO").expect("FOO not found");  // panicked at FOO not found: EnvVar(NotPresent)
```

Should `dotenvy::Error` contain the name of the env var somewhere?

Relevant
#14 
https://users.rust-lang.org/t/is-std-varerror-designed-like-this-on-purpose-and-if-so-why/53679

**Comment by sonro** [comment]
I suggest that the error be kept as is.

## Reason

The only place where the error can occur is the [`dotenvy::var`] function that @allan2 mentioned. This is a thin wrapper around `dotenv().ok()` and [`std::env::var`](https://doc.rust-lang.org/std/env/fn.var.html), which is probably why the author chose to emit the same error.

The forum post's [conclusion](https://users.rust-lang.org/t/is-std-varerror-designed-like-this-on-purpose-and-if-so-why/53679/3) about the underlying `VarError` seems appropriate for this crate. It is the expected behaviour of the standard library's `var` function, which this crate just wraps.

> Also, if you are not needing that information (maybe because you know the environment variable you were interested), it would make the error bigger and more expensive to create than necessary.

## Solution

Improving the documentation around this topic might be the best course of action. For example showing how to wrap the error using [`anyhow::Context`](https://docs.rs/anyhow/1.0.66/anyhow/trait.Context.html), mapping it onto a custom error type, or adding an `expect` message as in @allan2's example:

> ```rust
> dotenvy::var("FOO").expect("FOO not found");  // panicked at FOO not found: EnvVar(NotPresent)
> ```

## Alternative

Alternatively, or perhaps additionally, a more detailed error could be put behind a [Cargo feature](https://doc.rust-lang.org/cargo/reference/features.html). That way users could opt in to using an error similar to PR #15.

## Sidenote

Interestingly the [`dotenv!`](https://github.com/allan2/dotenvy/blob/a405018eb368b8ec92605157f40e4dc7c8d756ba/dotenv_codegen_impl/src/lib.rs#L56) macro uses the name of the missing variable in its error message. Again, it mimics the behviour of the standard library: [`env!`](https://doc.rust-lang.org/std/macro.env.html).

The [`dotenvy::var`] function only ever returns `Error::EnvVar`. The fact that it calls `dotenv().ok` is unrelated to its behaviour when it comes to checking a variable exists.

[`dotenvy::var`]: https://github.com/allan2/dotenvy/blob/a405018eb368b8ec92605157f40e4dc7c8d756ba/dotenv/src/lib.rs#L37

**Comment by allan2** [comment]
v0.16 now includes the path to the env file in the error message.
